litex/soc/software/bios/commands/cmd_boot.c

```diff
@@ -0,0 +1,107 @@
+#include <stdio.h>
+#include <stdlib.h>
+
+#include <generated/csr.h>
+
+#include "../command.h"
+#include "../helpers.h"
+#include "../boot.h"
+
+/**************************************
+ *            Flash boot              *
+ **************************************
+ */
+#ifdef FLASH_BOOT_ADDRESS
+static void flashboot_handler(int nb_params, char **params)
+{
+	flashboot();
+}
+
+struct command_struct cmd_flashboot =
+{
+	.func = flashboot_handler,
+	.name = "flashboot",
+	.help = "Boot from flash",
+};
+
+define_command(cmd_flashboot);
+#endif
+
+/**************************************
+ *              ROM boot              *
+ **************************************
+ */
+#ifdef ROM_BOOT_ADDRESS
+static void romboot_handler(int nb_params, char **params)
+{
+	romboot();
+}
+
+struct command_struct cmd_romboot =
+{
+	.func = romboot_handler,
+	.name = "romboot",
+	.help = "Boot from embedded rom",
+};
+
+define_command(cmd_romboot);
+#endif
+
+/**************************************
+ *            Serial boot             *
+ **************************************
+ */
+static void serialboot_handler(int nb_params, char **params)
+{
+	serialboot();
+}
+
+struct command_struct cmd_serialboot =
+{
+	.func = serialboot_handler,
+	.name = "serialboot",
+	.help = "Boot via SFL",
+};
+
+define_command(cmd_serialboot);
+
```

Comment:
I was wondering if the handler / struct / define_command call should all be one big macro (or maybe even autogenerated using a Python script)? Don't know if that is a good idea or not?

Reply:
A big macro could save us a lot of line code. Especially if make .func point directly to functions that don't require parameters.

Reply:
I'm not a *huge* fan of macro abuse, but it does seem like a lot of code could be saved.

The fact the bios is already part of a Python based build system might mean it is better to use a Python script to generate code rather than making the C preprocessor do the work?

Reply:
> I'm not a huge fan of macro abuse
Neither am I.

> a Python script to generate code

From my point of view, doing some auto-generated C code is like doing a big macro.
So, I would vote for a C macro here.

Reply:
@mithro: i'm not sure we should abuse too much of Python/C code mixing. It's probably fine to have auto-generated access functions or headers from Python, but not sure this is good idea for this part of the code, so i would also vote for C macros for this part (at least for now, we could change it in the future if we see if would really simplify things).

Reply:
I created a macro doing this:
https://github.com/fjullien/litex/blob/term_hist_auto_compl/litex/soc/software/bios/command.h#L36